### PR TITLE
Bluetooth: Host: Re-key service changed config on identity resolution

### PR DESCRIPTION
**Pull request by HaavardRei**
Fixes a bug where if a privacy-enabled client subscribes to Service Changed CCC before bonding, 
the SC config is initially stored under the peer's RPA. On  reconnection with a new RPA, sc_restore()
would look up the config by identity address and fails to find it, so the SC indication is never delivered.
The bug is fixed by re-keying the entry in the identity_resolved callback, similar to what is done for 
CCC and CF.

Adds a bsim test to verify the correct behavior.

### CODE REVIEW

File: tests/bsim/bluetooth/host/gatt/sc_indicate/test_scripts/sc_indicate_privacy.sh [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+# uses privacy (RPA) and subscribes to SC CCC before bonding. Before bonding the SC is
+# stored under the central's RPA. On bonding, the SC is stored under the identity address
+# through the bt_gatt_identity_resolved callback. This test will fail if the SC indication
+# is not delivered after reconnection.
+
+source ${ZEPHYR_BASE}/tests/bsim/sh_common.source
+
+test_name='sc_indicate_privacy'
+test_exe="bs_${BOARD_TS}_tests_bsim_bluetooth_host_gatt_${test_name}_overlay-privacy_conf"
+simulation_id="${test_name}"
```
Comment (aescolar) [inline tests/bsim/bluetooth/host/gatt/sc_indicate/test_scripts/sc_indicate_privacy.sh]:
```suggestion
test_exe="bs_${BOARD_TS}_tests_bsim_bluetooth_host_gatt_sc_indicate_overlay-privacy_conf"
simulation_id='sc_indicate_privacy'
```
Major: The executable name needs to match what is set in the yaml
Nit: There is no need to have an intermediate variable that is used only once

File: tests/bsim/bluetooth/host/gatt/sc_indicate/src/bs_bt_utils.h [hunk 2]
```diff
@@ -29,3 +29,4 @@ void stop_adv(struct bt_le_ext_adv *adv);
 void set_security(bt_security_t sec);
 void pairing_complete(struct bt_conn *conn, bool bonded);
 void pairing_failed(struct bt_conn *conn, enum bt_security_err err);
+extern bool privacy;
```
Comment (aescolar) [inline tests/bsim/bluetooth/host/gatt/sc_indicate/src/bs_bt_utils.h]: Nit: Although this works, it is quite ugly to have a global for this, specially if it is called just something as generic and likely to name clash as `privacy`.
It would be nicer to have either a more specific name, or better a function to get it and the variable be private.

**Review by aescolar — DISMISSED** [review]

File: tests/bsim/bluetooth/host/gatt/sc_indicate/test_scripts/sc_indicate_privacy.sh (hunk 1 above)
Comment (HaavardRei) [reply tests/bsim/bluetooth/host/gatt/sc_indicate/test_scripts/sc_indicate_privacy.sh]: Good point(s)! Added

File: tests/bsim/bluetooth/host/gatt/sc_indicate/src/bs_bt_utils.h (hunk 2 above)
Comment (HaavardRei) [reply tests/bsim/bluetooth/host/gatt/sc_indicate/src/bs_bt_utils.h]: Heh yeah agreed. Changed it to use a get function!

File: tests/bsim/bluetooth/host/gatt/sc_indicate/src/main.c [hunk 3]
```diff
@@ -4,20 +4,45 @@
...
+}
+
+static void test_args(int argc, char *argv[])
+{
+	bs_args_struct_t args_struct[] = {
+		{.dest = &privacy_enabled,
+		 .type = 'b',
+		 .name = "{0, 1}",
+		 .option = "privacy",
+		 .descript = "Use privacy (true) or not (false)."},
```
Comment (m-alperen-sener) [inline tests/bsim/bluetooth/host/gatt/sc_indicate/src/main.c]:
```suggestion
		 .descript = "Use privacy enabled or disabled."},
```

File: tests/bsim/bluetooth/host/gatt/sc_indicate/src/peripheral.c [hunk 4]
```diff
@@ -81,6 +82,11 @@ void peripheral(void)
 	TEST_ASSERT(!err, "bt_gatt_service_register failed (%d)", err);
 	LOG_DBG("New service added");
 
+	if (run_with_privacy()) {
+		/* Wait so that the central has time to rotate its RPA. */
+		k_sleep(K_SECONDS(3));
```
Comment (m-alperen-sener) [inline tests/bsim/bluetooth/host/gatt/sc_indicate/src/peripheral.c]: better to get this value from `BT_RPA_TIMEOUT` or if `BT_RPA_TIMEOUT_DYNAMIC` from `bt_dev.rpa_timeout` IMO

File: tests/bsim/bluetooth/host/gatt/sc_indicate/testcase.yaml [hunk 5]
```diff
@@ -8,3 +8,14 @@ tests:
...
       bsim_exe_name: tests_bsim_bluetooth_host_gatt_sc_indicate_prj_conf
+  bluetooth.host.gatt.sc_indicate_privacy:
+    build_only: true
+    tags:
+      - bluetooth
+    platform_allow:
+      - nrf52_bsim/native
+    harness: bsim
+    extra_args:
+      - EXTRA_CONF_FILE=overlay-privacy.conf
```
Comment (m-alperen-sener) [inline tests/bsim/bluetooth/host/gatt/sc_indicate/testcase.yaml:19]: only this test case uses PRIVACY and generating an executable anyway just for privacy enabled not using an existing executable!

so why do we include a bool state variable privacy and setting it with test argument: wouldn't it be easier to just replace `run_with_privacy(void)` with IS_ENABLED(CONFIG_BT_PRIVACY);

File: tests/bsim/bluetooth/host/gatt/sc_indicate/test_scripts/sc_indicate_privacy.sh [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+#
+# This test verifies that the Service Changed indication is delivered when the central
+# uses privacy (RPA) and subscribes to SC CCC before bonding. Before bonding the SC is
+# stored under the central's RPA. On bonding, the SC is stored under the identity address
+# through the bt_gatt_identity_resolved callback. This test will fail if the SC indication
+# is not delivered after reconnection.
+
+source ${ZEPHYR_BASE}/tests/bsim/sh_common.source
+
+test_exe="bs_${BOARD_TS}_tests_bsim_bluetooth_host_gatt_sc_indicate_overlay-privacy_conf"
```
Comment (m-alperen-sener) [inline tests/bsim/bluetooth/host/gatt/sc_indicate/test_scripts/sc_indicate_privacy.sh:13]: is it possible to use this executable with privacy=0?

**Review by m-alperen-sener — CHANGES_REQUESTED** [review]

File: tests/bsim/bluetooth/host/gatt/sc_indicate/testcase.yaml (hunk 5 above)
Comment (HaavardRei) [reply tests/bsim/bluetooth/host/gatt/sc_indicate/testcase.yaml:19]: True, this is indeed easier. Added.

File: tests/bsim/bluetooth/host/gatt/sc_indicate/src/peripheral.c (hunk 4 above)
Comment (HaavardRei) [reply tests/bsim/bluetooth/host/gatt/sc_indicate/src/peripheral.c]: CONFIG_BT_RPA_TIMEOUT is undefined for the non-privacy test case and will give a compile error. I added a commend above each sleep, alternatively we could redefine it using some #if gymnastics if this is not sufficient.

File: tests/bsim/bluetooth/host/gatt/sc_indicate/test_scripts/sc_indicate_privacy.sh (hunk 6 above)
Comment (HaavardRei) [reply tests/bsim/bluetooth/host/gatt/sc_indicate/test_scripts/sc_indicate_privacy.sh:13]: Not anymore :)

File: tests/bsim/bluetooth/host/gatt/sc_indicate/src/peripheral.c (hunk 4 above)
Comment (m-alperen-sener) [reply tests/bsim/bluetooth/host/gatt/sc_indicate/src/peripheral.c]:
```suggestion
#if defined(CONFIG_BT_PRIVACY)
	k_sleep(K_SECONDS(CONFIG_BT_RPA_TIMEOUT + 2));
#endif
```

yeap I guess this requires #ifdef gymnastic;

Comment (HaavardRei) [reply tests/bsim/bluetooth/host/gatt/sc_indicate/src/peripheral.c]: Added.

File: tests/bsim/bluetooth/host/gatt/sc_indicate/src/central.c [hunk 7]
```diff
@@ -175,19 +175,36 @@ void central(void)
...
+	if (!IS_ENABLED(CONFIG_BT_PRIVACY)) {
+		gatt_discover();
+		subscribe();
+	}
 
 	disconnect();
 	wait_disconnected();
 	clear_g_conn();
 
+#if IS_ENABLED(CONFIG_BT_PRIVACY) && IS_ENABLED(CONFIG_BT_RPA_TIMEOUT)
```
Comment (m-alperen-sener) [inline tests/bsim/bluetooth/host/gatt/sc_indicate/src/central.c]: `CONFIG_BT_RPA_TIMEOUT` is already depends on `CONFIG_BT_PRIVACY` also I don't see why we are forcing IS_ENABLED here instead of #ifdef or #if defined; this is a test app not a library or stack code for production?

I would like to know if there is a benefit in doing so.

File: tests/bsim/bluetooth/host/gatt/sc_indicate/src/peripheral.c [hunk 8]
```diff
@@ -81,6 +82,11 @@ void peripheral(void)
 	TEST_ASSERT(!err, "bt_gatt_service_register failed (%d)", err);
 	LOG_DBG("New service added");
 
+#if IS_ENABLED(CONFIG_BT_PRIVACY) && IS_ENABLED(CONFIG_BT_RPA_TIMEOUT)
```
Comment (m-alperen-sener) [inline tests/bsim/bluetooth/host/gatt/sc_indicate/src/peripheral.c]: same here.

**Review by m-alperen-sener — CHANGES_REQUESTED** [review]

**Review by m-alperen-sener — APPROVED** [review]
thanks 👍

File: subsys/bluetooth/host/gatt.c [hunk 9]
```diff
@@ -1031,6 +1031,30 @@ static void bt_gatt_identity_resolved(struct bt_conn *conn, const bt_addr_le_t *
 			bt_gatt_store_cf(conn->id, &conn->le.dst);
 		}
 	}
+
+#if defined(CONFIG_BT_GATT_SERVICE_CHANGED)
```
Comment (Thalley) [inline subsys/bluetooth/host/gatt.c]: Is anything us from using `IS_ENABLED` here instead of `#if defined`?

File: subsys/bluetooth/host/gatt.c [hunk 10]
```diff
@@ -1031,6 +1031,30 @@ static void bt_gatt_identity_resolved(struct bt_conn *conn, const bt_addr_le_t *
...
+	/* Keep Service Changed bookkeeping aligned with the bonded identity
+	 * address so restore-on-reconnect works when privacy is enabled.
+	 */
+	struct gatt_sc_cfg *sc = find_sc_cfg(conn->id, private_addr);
+
+	if (sc) {
```
Comment (Thalley) [inline subsys/bluetooth/host/gatt.c]:
```suggestion
	if (sc != NULL) {
```

File: subsys/bluetooth/host/gatt.c [hunk 11]
```diff
@@ -1031,6 +1031,30 @@ static void bt_gatt_identity_resolved(struct bt_conn *conn, const bt_addr_le_t *
...
+		struct gatt_sc_cfg *sc_id = find_sc_cfg(conn->id, id_addr);
+
+		if (sc_id && sc_id != sc) {
```
Comment (Thalley) [inline subsys/bluetooth/host/gatt.c]:
```suggestion
		if (sc_id != NULL && sc_id != sc) {
```

File: tests/bsim/bluetooth/host/gatt/sc_indicate/src/central.c [hunk 12]
```diff
@@ -175,19 +175,36 @@ void central(void)
...
+	if (!IS_ENABLED(CONFIG_BT_PRIVACY)) {
+		gatt_discover();
+		subscribe();
+	}
 
 	disconnect();
 	wait_disconnected();
 	clear_g_conn();
 
+#if defined(CONFIG_BT_PRIVACY)
```
Comment (Thalley) [inline tests/bsim/bluetooth/host/gatt/sc_indicate/src/central.c:194]: `IS_ENABLED`?

File: tests/bsim/bluetooth/host/gatt/sc_indicate/src/peripheral.c [hunk 13]
```diff
@@ -81,6 +82,11 @@ void peripheral(void)
 	TEST_ASSERT(!err, "bt_gatt_service_register failed (%d)", err);
 	LOG_DBG("New service added");
 
+#if defined(CONFIG_BT_PRIVACY)
```
Comment (Thalley) [inline tests/bsim/bluetooth/host/gatt/sc_indicate/src/peripheral.c:85]: `IS_ENABLED`?

File: tests/bsim/bluetooth/host/gatt/sc_indicate/src/central.c [hunk 14]
```diff
@@ -175,19 +175,36 @@ void central(void)
 	scan_connect_to_first_result();
 	wait_connected();
 
+	if (IS_ENABLED(CONFIG_BT_PRIVACY)) {
+		/* Subscribe to SC CCC before bonding. With privacy enabled
+		 * conn->le.dst is still the central's RPA, so the peripheral
+		 * stores the SC config under that RPA.
```
Comment (Thalley) [inline tests/bsim/bluetooth/host/gatt/sc_indicate/src/central.c]: Do we need to wait until after `set_security(BT_SECURITY_L2);` for the non-Privacy case, or could we just always do discover before pairing?

Comment (Thalley) [inline tests/bsim/bluetooth/host/gatt/sc_indicate/src/central.c]: If `CONFIG_BT_ATT_RETRY_ON_SEC_ERR=y` (default) and there's anything that requires encryption, then pairing is done during discovery/subscribing. For the comment to be true, we need to ensure that we aren't pairing during discovery and/or subscribing

File: tests/bsim/bluetooth/host/gatt/sc_indicate/src/central.c [hunk 15]
```diff
@@ -175,19 +175,36 @@ void central(void)
 	scan_connect_to_first_result();
 	wait_connected();
 
+	if (IS_ENABLED(CONFIG_BT_PRIVACY)) {
+		/* Subscribe to SC CCC before bonding. With privacy enabled
```
Comment (Thalley) [inline tests/bsim/bluetooth/host/gatt/sc_indicate/src/central.c]:
```suggestion
		/* Subscribe to SC CCC before pairing. With privacy enabled
```

Resolving addresses can be done during pairing without bonding too, right?

File: subsys/bluetooth/host/gatt.c [hunk 16]
```diff
@@ -1031,6 +1031,30 @@ static void bt_gatt_identity_resolved(struct bt_conn *conn, const bt_addr_le_t *
...
+					   sc->data.end);
+			clear_sc_cfg(sc);
+			sc = sc_id;
+		} else {
+			bt_addr_le_copy(&sc->peer, id_addr);
+		}
+
+		if (is_bonded) {
+			sc_store(sc);
+		}
```
[the remark this answers is not in the document]
Comment (HaavardRei) [reply subsys/bluetooth/host/gatt.c]: The entry is previously only stored here besides the reset: 
https://github.com/zephyrproject-rtos/zephyr/blob/acc53653b4707aa40c22aec8c33f1d733e5e670f/subsys/bluetooth/host/gatt.c#L321-L323
Where we look up whether a bond exists. Thus I do not believe we ever persist SC entries using RPA.

[the remark this answers is not in the document]
Comment (Thalley) [reply subsys/bluetooth/host/gatt.c]: Hmm, I wonder if we do? 

https://github.com/zephyrproject-rtos/zephyr/issues/96614

If we connect to a device that uses an NRPA but where we still do bonding, the we never "resolve" the NRPA to the identity address, and `conn->ls.dst` remains the NRPA I think. 

^ is a corner case that should not be handled by GATT anyways, but I think this may technically happen until that is fixed, unless I'm mistaking this comment from Copilot with something else

File: tests/bsim/bluetooth/host/gatt/sc_indicate/src/central.c (hunk 12 above)
Comment (HaavardRei) [reply tests/bsim/bluetooth/host/gatt/sc_indicate/src/central.c:194]: Will give a compilation error for the non-privacy testcase as CONFIG_BT_RPA_TIMEOUT is not defined.

File: subsys/bluetooth/host/gatt.c (hunk 9 above)
Comment (HaavardRei) [reply subsys/bluetooth/host/gatt.c]: Don't think so, I'll add it!

File: tests/bsim/bluetooth/host/gatt/sc_indicate/src/peripheral.c (hunk 13 above)
Comment (HaavardRei) [reply tests/bsim/bluetooth/host/gatt/sc_indicate/src/peripheral.c:85]: Same as above.

File: tests/bsim/bluetooth/host/gatt/sc_indicate/src/central.c (hunk 14 above)
Comment (HaavardRei) [reply tests/bsim/bluetooth/host/gatt/sc_indicate/src/central.c]: I don't think doing so will affect what the non-privacy testcase tests, so I'll add it :)

Comment (HaavardRei) [reply tests/bsim/bluetooth/host/gatt/sc_indicate/src/central.c]: True, I added the config to make sure.

**Review by Thalley — DISMISSED** [review]

**Review by Thalley — DISMISSED** [review]

**Review by m-alperen-sener — APPROVED** [review]

File: subsys/bluetooth/host/gatt.c
Comment (Thalley) [inline subsys/bluetooth/host/gatt.c:1]: Outside the scope of this PR I think, but `bt_gatt_identity_resolved` is called in the BT RX thread and it modifies the content of `gatt_sc_cfg`. The `gatt_sc_cfg`s are also accessed by the system workqueue (e.g. via `do_db_hash->set_all_change_unaware`) or from any other thread via `bt_gatt_notify_cb->gatt_notify->bt_gatt_change_aware`

File: subsys/bluetooth/host/gatt.c [hunk 17]
```diff
@@ -1031,6 +1031,30 @@ static void bt_gatt_identity_resolved(struct bt_conn *conn, const bt_addr_le_t *
...
+		 * address so restore-on-reconnect works when privacy is enabled.
+		 */
+		struct gatt_sc_cfg *sc = find_sc_cfg(conn->id, private_addr);
+
+		if (sc != NULL) {
+			struct gatt_sc_cfg *sc_id = find_sc_cfg(conn->id, id_addr);
+
+			if ((sc_id != NULL) && (sc_id != sc)) {
+				(void)update_range(&sc_id->data.start, &sc_id->data.end,
+						   sc->data.start, sc->data.end);
```
[the remark this answers is not in the document]
Comment (HaavardRei) [reply subsys/bluetooth/host/gatt.c]: I do not think this is the place to verify the correctness of this struct (might be done in find_sc_cfg instead?) or if this is needed.

File: subsys/bluetooth/host/gatt.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (HaavardRei) [reply subsys/bluetooth/host/gatt.c]: Hmm I think you're right. In which case we may get orphaned entries if new NRPAs are used.

File: subsys/bluetooth/host/gatt.c (hunk 17 above)
[the remark this answers is not in the document]
Comment (KyraLengfeld) [reply subsys/bluetooth/host/gatt.c]: Just looking at this code section, I think copilot has a point. Question is whether at this point in the code sc_id would ever be zero when sc is not? If we can confidently say that isn't a possibility, then we don't need to guard against it. (Need to familiarize myself with sc struct and how it is populated and updated still though, but you might know already :) )

[the remark this answers is not in the document]
Comment (HaavardRei) [reply subsys/bluetooth/host/gatt.c]: Could instead add a simpler variant like:
```
if (sc->data.start || sc->data.end) {
     (void)update_range(&sc_id->data.start, &sc_id->data.end,
                       sc->data.start, sc->data.end);
}
```
instead of what Copilot suggests? I think this covers the case it refers to.

[the remark this answers is not in the document]
Comment (KyraLengfeld) [reply subsys/bluetooth/host/gatt.c]: You probably found that as well, but this is also in line with the rest of the code, like:
https://github.com/zephyrproject-rtos/zephyr/blob/b1348f3f50a5c4c92d90f4d284df32c2b9bfa9d3/subsys/bluetooth/host/gatt.c#L3175
so think this is a reasonable check to add, otherwise I think this PR looks good.

**Review by KyraLengfeld — DISMISSED** [review]

File: subsys/bluetooth/host/gatt.c [hunk 18]
```diff
@@ -1031,6 +1031,32 @@ static void bt_gatt_identity_resolved(struct bt_conn *conn, const bt_addr_le_t *
...
+		/* Keep Service Changed bookkeeping aligned with the bonded identity
+		 * address so restore-on-reconnect works when privacy is enabled.
+		 */
+		struct gatt_sc_cfg *sc = find_sc_cfg(conn->id, private_addr);
+
+		if (sc != NULL) {
+			struct gatt_sc_cfg *sc_id = find_sc_cfg(conn->id, id_addr);
+
+			if ((sc_id != NULL) && (sc_id != sc)) {
+				if (sc->data.start || sc->data.end) {
```
Comment (Thalley) [inline subsys/bluetooth/host/gatt.c]:
```suggestion
				if (sc->data.start != 0U || sc->data.end != 0U) {
```

Comment (Thalley) [inline subsys/bluetooth/host/gatt.c]: Does `sc->data.start` need to be `<= sc->data.end`? 

Looking at `update_range` it does not check this either, and it looks like `data.end` can end up smaller than `data.start` which seems weird to me

Comment (HaavardRei) [reply subsys/bluetooth/host/gatt.c]: Can we have start=end? If so, I propose adding a early check to `update_range` like  `if (new_start > new_end) ...` --> abort+LOG_ERR.

Comment (Thalley) [reply subsys/bluetooth/host/gatt.c]: I'm not too familiar with this part of GATT, so I don't know what the right solution is :) If it's exclusively a local value check that should never happen, then perhaps an assert? We can also do LOG_ERR if you prefer that

> Can we have start=end?

No idea :)

Comment (HaavardRei) [reply subsys/bluetooth/host/gatt.c]: Added!

Comment (HaavardRei) [reply subsys/bluetooth/host/gatt.c]: Added an assert. I believe it could be possible to have a single attribute (just declaration, no characteristics) so the assert allows start=end :)

**Review by Thalley — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Review by KyraLengfeld — APPROVED** [review]
